### PR TITLE
tversky loss 追加

### PR DESCRIPTION
**Pull request by Yuto-web**
## 変更点

- tvelsky lossを追加
    - 参考リンク：http://ni4muraano.hatenablog.com/entry/2018/04/02/223236

## 問題点

tvelsky_lossのパラメーター(alpha, beta)がyamlで変更できない（他のlossも同様）

## 備考

- farmerが実行できることは確認済み
- 学習がうまく進むかは確認中

### CODE REVIEW

File: farmer/ncc/losses/custom_loss.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from keras import backend as K
+import tensorflow as tf
+
+def tversky_index(y_true, y_pred, alpha=0.3, beta=0.7):
+    # 参考サイト：http://ni4muraano.hatenablog.com/entry/2018/04/02/223236
+
+    y_true = K.flatten(y_true)
+    y_pred = K.flatten(y_pred)
+    intersection = K.sum(y_true * y_pred)
```
Comment (aiorhiroki) [inline farmer/ncc/losses/custom_loss.py:9]: true_positive派だけど、みんなは何派？

File: farmer/ncc/losses/custom_loss.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    false_positive = K.sum((1.0 - y_true) * y_pred)
+    false_negative = K.sum(y_true * (1.0 - y_pred))
+    return intersection / (intersection + alpha*false_positive + beta*false_negative)
```
Comment (aiorhiroki) [inline farmer/ncc/losses/custom_loss.py:12]: true_negativeだけの画像の時に、ゼロ割のリスクってあるかな？

File: farmer/domain/tasks/build_model_task.py [hunk 3]
```diff
@@ -6,6 +6,8 @@
 )
 
 from farmer.ncc.models import xception, mobilenet, Deeplabv3, Model2D
+# from farmer.ncc.losses.custom_loss import tversky_loss
```
Comment (aiorhiroki) [inline farmer/domain/tasks/build_model_task.py:9]: このコメントは消して良さそう

File: farmer/ncc/losses/custom_loss.py (hunk 1 above)
Comment (yumion) [reply farmer/ncc/losses/custom_loss.py:9]: 下に合わせるとtrue_positiveですね
普段はtp, fp, fn派です

File: farmer/ncc/losses/custom_loss.py (hunk 2 above)
Comment (yumion) [reply farmer/ncc/losses/custom_loss.py:12]: eps=1e-7を引数で定義して、分母に足してあげたほうがいいですね

File: farmer/ncc/losses/custom_loss.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from keras import backend as K
+import tensorflow as tf
+
+def tversky_index(y_true, y_pred, alpha=0.3, beta=0.7):
```
Comment (aiorhiroki) [inline farmer/ncc/losses/custom_loss.py:4]:
```
def tversky_loss(alpha=0.3, beta=0.7):
    def tversky_index(y_true, y_pred, alpha, beta):
        y_true = K.flatten(y_true)
        y_pred = K.flatten(y_pred)
        intersection = K.sum(y_true * y_pred)
        false_positive = K.sum((1.0 - y_true) * y_pred)
        false_negative = K.sum(y_true * (1.0 - y_pred))
        return intersection / (intersection + alpha*false_positive + beta*false_negative)
    def loss(y_true, y_pred):
        y_pred = tf.clip_by_value(
            y_pred, tf.keras.backend.epsilon(), 1 - tf.keras.backend.epsilon())
        loss =tversky_index(y_true, y_pred, alpha, beta)
        return tf.reduce_mean(loss)
    return loss
```

Comment (Yuto-web) [reply farmer/ncc/losses/custom_loss.py:4]: ありがとうございます。
書き直します。

File: farmer/ncc/losses/custom_loss.py (hunk 1 above)
Comment (Yuto-web) [reply farmer/ncc/losses/custom_loss.py:9]: ありがとうございます。
true_positiveに修正します。

File: farmer/ncc/losses/custom_loss.py (hunk 2 above)
Comment (Yuto-web) [reply farmer/ncc/losses/custom_loss.py:12]: ありがとうございます。
tversky_loss 内のグローバル変数としてeps = tf.keras.backend.epsilon()してepsをinnner関数で使うようにしました。

Comment (yumion) [reply farmer/ncc/losses/custom_loss.py:12]: epsは分母分子両方に足すみたいです

File: farmer/domain/tasks/build_model_task.py [hunk 5]
```diff
@@ -104,6 +106,12 @@ def _do_make_model_task(
                     input_shape=(height, width, 3),
                     classes=nb_classes,
                 )
+            elif model_name == "FPN":
```
Comment (yumion) [inline farmer/domain/tasks/build_model_task.py:109]: https://github.com/aiorhiroki/farmer/pull/96
これと被っているので分けましょう

File: farmer/ncc/losses/custom_loss.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from keras import backend
```
Comment (yumion) [inline farmer/ncc/losses/custom_loss.py:1]: 慣例的に
```
from keras import backend as K
```
です

File: farmer/domain/tasks/build_model_task.py [hunk 7]
```diff
@@ -170,12 +171,21 @@ def _do_compile_model_task(
                 print('------------------')
                 print('Loss:', loss_func)
                 print('------------------')
-                model.compile(
-                    optimizer=optimizer,
-                    loss=globals()[loss_func],
-                    metrics=[metrics.iou_score,
-                             categorical_crossentropy],
-                )
+                if loss_func == 'tversky_loss':
```
Comment (yumion) [inline farmer/domain/tasks/build_model_task.py:174]: これif文にしなくても `globals()[loss_func]` でtversky_loss取ってこれますよ

File: farmer/domain/tasks/build_model_task.py [hunk 8]
```diff
@@ -170,12 +171,21 @@ def _do_compile_model_task(
...
+                    model.compile(
+                        optimizer=optimizer,
+                        loss=loss,
+                        metrics=[metrics.iou_score,
+                                categorical_crossentropy],
+                    )
+                else:
+                    model.compile(
+                        optimizer=optimizer,
+                        loss=globals()[loss_func],
```
Comment (yumion) [inline farmer/domain/tasks/build_model_task.py:185]: alpha, betaのパラメータは
```
loss=globals()[loss_func](**config.loss_params),
```
とかやれると綺麗かなと(globals()[loss_func]で引数入れられるのかは知らないですが)

File: farmer/ncc/losses/custom_loss.py (hunk 6 above)
Comment (Yuto-web) [reply farmer/ncc/losses/custom_loss.py:1]: ありがとうございます。修正しました

File: farmer/domain/tasks/build_model_task.py (hunk 7 above)
Comment (Yuto-web) [reply farmer/domain/tasks/build_model_task.py]: loss functionsのリファクタリングを反映し、lossは全てgetattrでとって来るようにしました

File: farmer/domain/tasks/build_model_task.py (hunk 8 above)
Comment (Yuto-web) [reply farmer/domain/tasks/build_model_task.py]: 上記と同じく、loss functionsのリファクタリングを反映し、lossは全てgetattrでとって来るようにしました

File: farmer/ncc/losses/custom_loss.py (hunk 6 above)
Comment (Yuto-web) [reply farmer/ncc/losses/custom_loss.py]: kerasを消してtfのみ使うようにしました。
